clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java

```diff
@@ -369,6 +390,78 @@ public void logUnused() {
+     */
+    @SuppressWarnings("unchecked")
+    private  Map<String, ?> resolveConfigVariables(Map<String, String> indirectVariables, Map<?, ?> configProviders, Map<String, Object> original) {
+        Map<String, ConfigProvider> providers;
+
+        if (configProviders == null || configProviders.isEmpty()) {
+            providers = instantiateConfigProvider(indirectVariables, original);
+        } else {
+
+            for (Map.Entry<?, ?> entry : configProviders.entrySet()) {
+                if (!(entry.getKey() instanceof String))
+                    configProviders.remove(entry.getKey());
+            }
+            providers = instantiateConfigProvider((Map<String, String>) configProviders, original);
+        }
+        if (!providers.isEmpty()) {
+            ConfigTransformer configTransformer = new ConfigTransformer(providers);
+            ConfigTransformerResult result = configTransformer.transform(indirectVariables);
+            original.putAll(result.data());
+        }
+
+        return original;
+    }
+
+    private Map<String, Object> getClassConfigProperties(String prefix, Map<String, ?> originals) {
+        Map<String, Object> result = new HashMap<>();
+        for (Map.Entry<String, ?> entry : originals.entrySet()) {
+            if (entry.getKey().startsWith(prefix) && entry.getKey().length() > prefix.length()) {
+                result.put(entry.getKey().substring(prefix.length()), entry.getValue());
+            }
+        }
+        return result;
+    }
+
+    private Map<String, ConfigProvider> instantiateConfigProvider(Map<String, String> indirectConfigs, Map<String, Object> originals) {
+        final String configProviders = indirectConfigs.get(CONFIG_PROVIDERS_CONFIG);
+
+        if (configProviders == null || configProviders.isEmpty())
+            return Collections.emptyMap();
+
+        Map<String, String> providerMap = new HashMap<>();
+
+        for (String provider: configProviders.split(",")) {
+            String providerClass = CONFIG_PROVIDERS_CONFIG + "." + provider + ".class";
+            if (indirectConfigs.containsKey(providerClass))
+                providerMap.put(provider, indirectConfigs.get(providerClass));
+
+        }
+        // Instantiate Config Providers
+        Map<String, ConfigProvider> configProviderInstances = new HashMap<String, ConfigProvider>();
+        for (Map.Entry<String, String> entry : providerMap.entrySet()) {
+            try {
+                String prefix = CONFIG_PROVIDERS_CONFIG + "." + entry.getKey() + ".";
+                Map<String, ?> configProperties = getClassConfigProperties(prefix, originals);
+                ConfigProvider provider = Utils.newInstance(entry.getValue(), ConfigProvider.class);
+                provider.configure(configProperties);
+                configProviderInstances.put(entry.getKey(), provider);
+            } catch (ClassNotFoundException e) {
+                log.warn("Failed to initialize the class:" + entry.getValue());
```

Comment:
Shouldn't we throw an exception since the config is not valid?

Reply:
If the configprovider is not valid or not specified we will skip the variable substitution step and will simply validate and parse the supplied configuration

Reply:
Then the log message should state the reason (e.g., the class was not found) and that the named provider will not be used.

However, it does seem strange that this is technically an invalid configuration, so why would we not just throw a ConfigException? Note around lines 723 how any configuration property whose value is a Class name will be invalid if a `ClassNotFoundException` is caught, and this leads to a `ConfigException`. Silently ignoring seems at best inconsistent but at worst a bad idea.